### PR TITLE
Generalize check-model in NonLinearExtension for quadratic equations

### PR DESCRIPTION
**Pull request by ajreynol**
This PR generalizes the check model algorithm that uses intervals in the NonLinearExtension to handle approximations for solutions to univariate quadratic equations.

It also updates the NonLinearExtension class with a technique to produce approximately accurate models when produceModels() is true, by asserting all intervals in the model as assertions and rechecking.

### CODE REVIEW

**Comment by ajreynol** [comment]
This gains approximately +140 SAT on QF_NRA: https://www.starexec.org/starexec/secure/details/job.jsp?id=27812.

**Comment by ajreynol** [comment]
Work towards #1450.

File: src/theory/arith/nonlinear_extension.h [hunk 1]
```diff
@@ -631,6 +684,15 @@ class NonlinearExtension {
...
   Node getApproximateConstant(Node c, bool isLower, unsigned prec) const;
+  /** get approximate sqrt
+   *
+   * This approximates the square root of positive constant c. If this method
+   * returns true, then l and u are updated to constants such that
+   *   l <= sqrt( c ) <= u
+   * The argument iter is the number of iterations in the binary search to
+   * perform.
+   */
+  bool getApproximateSqrt(Node c, Node& l, Node& u, unsigned iter = 15) const;
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: comment on the magic number

File: src/theory/arith/nonlinear_extension.cpp [hunk 2]
```diff
@@ -978,257 +915,440 @@ bool NonlinearExtension::checkModelTf(const std::vector<Node>& assertions)
...
-          }
-          Node vbound = vbs.size() == 1 ? vbs[0] : nm->mkNode(MULT, vbs);
-          sum_bound.push_back(ArithMSum::mkCoeffTerm(m.second, vbound));
-        }
+        d_check_model_solved[eq] = Node::null();
+        return true;
       }
-      Node bound;
-      if (sum_bound.size() > 1)
+      else
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: no-else-after-return

File: src/theory/arith/nonlinear_extension.cpp [hunk 3]
```diff
@@ -978,257 +915,440 @@ bool NonlinearExtension::checkModelTf(const std::vector<Node>& assertions)
...
+        addCheckModelSubstitution(uvf, uvfv);
+        // recurse
+        return solveEqualitySimple(eq);
+      }
+    }
+    Trace("nl-ext-cms") << "...fail due to constrained invalid terms."
+                        << std::endl;
+    return false;
+  }
+  else if (var.isNull() || var.getType().isInteger())
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.cpp]: ditto

Comment (ajreynol) [reply src/theory/arith/nonlinear_extension.cpp]: This else doesn't cost any space, i prefer having these.

File: src/theory/arith/nonlinear_extension.h [hunk 4]
```diff
@@ -690,7 +690,9 @@ class NonlinearExtension {
    * returns true, then l and u are updated to constants such that
    *   l <= sqrt( c ) <= u
    * The argument iter is the number of iterations in the binary search to
-   * perform.
+   * perform. By default, this is set ot 15, which is usually enough to be
```
Comment (HanielB) [inline src/theory/arith/nonlinear_extension.h]: ot --> to

**Pull request merged**